llvm/lib/Target/AMDGPU/MCTargetDesc/AMDGPUMCExpr.h

```diff
@@ -0,0 +1,73 @@
+//===- AMDGPUMCExpr.h - AMDGPU specific MC expression classes ---*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_LIB_TARGET_AMDGPU_MCTARGETDESC_AMDGPUMCEXPR_H
+#define LLVM_LIB_TARGET_AMDGPU_MCTARGETDESC_AMDGPUMCEXPR_H
+
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/MC/MCExpr.h"
+
+namespace llvm {
+
+/// AMDGPU target specific variadic MCExpr operations.
+///
+/// Takes in a minimum of 1 argument to be used with an operation. The supported
+/// operations are:
+///   - (logic) or
+///   - max
+///
+/// \note If the 'or'/'max' operations are provided only a single argument, the
+/// operation will act as a no-op and simply resolve as the provided argument.
+///
+class AMDGPUVariadicMCExpr : public MCTargetExpr {
+public:
+  enum AMDGPUVariadicKind { AGVK_None, AGVK_Or, AGVK_Max };
```

Comment:
tiny nit: you can remove the `AMDGPU` prefix here as it's already contained in `AMDGPUVariadicMCExpr`

also just noticed this: we don't need the `None` kind if it's not supported. I think we'll `unreachable` if we try to evaluate a none.
If you want to keep the none kind add an assert that `Kind != AGVK_None` in the constructor so we don't accidentally construct those

Reply:
WRT None: I went with the latter option of adding an assert in the constructor as I've been using `AGVK_None` in the AMDGPUAsmParser but do let me know if the preference is to remove it altogether. I'm sure I can find a way to parse without it.